internal/tasks/image_gc.go

```diff
@@ -227,25 +231,25 @@ func (j *Janitor) evaluatePolicy(ctx context.Context, proc *processor.Processor,
 
 		// track matching "delete" policies in GCStatus to allow users insight
 		// into how policies match
-		if policy.Action == "delete" {
-			m.GCStatus.RelevantPolicies = append(m.GCStatus.RelevantPolicies, policy)
+		if gcPolicies.Action == "delete" {
+			m.GCStatus.RelevantPolicies = append(m.GCStatus.RelevantPolicies, gcPolicies)
 		}
 
 		// evaluate constraints
-		if !policy.MatchesTags(m.TagNames) {
+		if !gcPolicies.MatchesTags(m.TagNames) {
 			continue
 		}
-		if !policy.MatchesTimeConstraint(m.Manifest, aliveManifests, j.timeNow()) {
+		if !gcPolicies.MatchesTimeConstraint(m.Manifest, aliveManifests, j.timeNow()) {
 			continue
 		}
 
-		pCopied := policy
+		pCopied := gcPolicies
 		// execute policy action
-		switch policy.Action {
+		switch gcPolicies.Action {
 		case "protect":
 			m.GCStatus.ProtectedByPolicy = &pCopied
 		case "delete":
-			err := proc.DeleteManifest(ctx, account, repo, m.Manifest.Digest, keppel.AuditContext{
+			err := proc.DeleteManifest(ctx, account, repo, m.Manifest.Digest, tagPolicies, keppel.AuditContext{
```

Comment:
This deletion should not be attempted if the tag policies forbid it. Otherwise image GC will get stuck on a useless error.

Instead, the manifest's GCStatus should gain a new field `ProtectedByTagPolicy` that gets filled in this case (which also means API visibility for this mechanism).

To avoid duplicating the tag policy evaluation between here and Processor.DeleteManifest(), I recommend returning a defined error type from DeleteManifest() if a tag policy blocks deletion:

```
type DeleteManifestBlockedByTagPolicyError struct {
  Policy keppel.TagPolicy
}
```

and checking for this here via `errext.As`.